corehq/apps/app_manager/fixtures/mobile_ucr.py

```diff
@@ -484,17 +381,126 @@ def _utcnow():
+    all_filter_values = {
+        filter_slug: restore_user.get_ucr_filter_value(filter, report.get_ui_filter(filter_slug))
+        for filter_slug, filter in report_config.filters.items()
+    }
+    # apply all prefilters
+    prefilters = [ReportFilterFactory.from_spec(p, report) for p in report.prefilters]
+    prefilter_values = {prefilter.name: prefilter.value() for prefilter in prefilters}
+    all_filter_values.update(prefilter_values)
+    # filter out nulls
+    filter_values = {
+        filter_slug: filter_value for filter_slug, filter_value in all_filter_values.items()
+        if filter_value is not None
+    }
+    defer_filters = [
+        report.get_ui_filter(filter_slug)
+        for filter_slug, filter_value in all_filter_values.items()
+        if filter_value is None and is_valid_mobile_select_filter_type(report.get_ui_filter(filter_slug))
+    ]
+    data_source.set_filter_values(filter_values)
+    data_source.set_defer_fields([f.field for f in defer_filters])
+    filter_options_by_field = defaultdict(set)
+
+    row_elements = get_report_element(
+        report_data_cache,
+        report_config,
+        data_source,
+        {f.field for f in defer_filters},
+        filter_options_by_field,
+        row_to_element,
+    )
+    filters_elem = _get_filters_elem(defer_filters, filter_options_by_field, restore_user._couch_user)
+
+    if (report_config.report_id in settings.UCR_COMPARISONS and
+        COMPARE_UCR_REPORTS.enabled(uuid.uuid4().hex, NAMESPACE_OTHER)):
+        compare_ucr_dbs.delay(domain, report_config.report_id, filter_values)
+
+    return row_elements, filters_elem
+
+
+def get_report_element(report_data_cache, report_config, data_source, deferred_fields, filter_options_by_field, row_to_element):
+    """
+    :param row_to_element: function (
+                deferred_fields, filter_options_by_field, row, index, is_total_row
+            ) -> row_element
+    """
+    if data_source.has_total_row:
+        total_row_calculator = IterativeTotalRowCalculator(data_source)
+    else:
+        total_row_calculator = MockTotalRowCalculator()
+
+    row_elements = []
+    row_index = 0
+    rows = report_data_cache.get_data(report_config.uuid, data_source)
+    for row_index, row in enumerate(rows):
+        row_elements.append(row_to_element(deferred_fields, filter_options_by_field, row, row_index))
+        total_row_calculator.update_totals(row)
+
+    if data_source.has_total_row:
+        total_row = total_row_calculator.get_total_row()
+        row_elements.append(row_to_element(
+            deferred_fields, filter_options_by_field,
+            dict(
+                zip(
+                    list(data_source.final_column_ids),
+                    list(map(str, total_row))
+                )
+            ),
+            row_index + 1,
+            is_total_row=True,
+        ))
+    return row_elements
+
+
+def _get_filters_elem(defer_filters, filter_options_by_field, couch_user):
```

Comment:
Could be worth adding a sample output as comment here to understand the code better.

Reply:
Non of the the other methods have docs about the structure of the returned data. There are test cases which validate the output e.g. https://github.com/dimagi/commcare-hq/blob/42aa55a1df620fb38fb9cf64b926a0733cd3700c/corehq/apps/app_manager/tests/test_report_config.py#L361

I agree the Mobile UCR docs aren't great but I don't think docstrings is the right place to document the XML structure.